protocol/src/format.rs

```diff
@@ -104,322 +253,444 @@ impl Component {
-                                'o' => modifier.italic = Some(true),
-                                'r' => {}
-                                _ => println!(
-                                    "warning: unsupported color code {:?} in text '{}'",
-                                    color_char, txt
-                                ),
-                            }
-
-                            current.modifier = modifier;
-                        }
-                    }
+            let red = u8::from_str_radix(&without_prefix[0..2], 16)?;
+            let green = u8::from_str_radix(&without_prefix[2..4], 16)?;
+            let blue = u8::from_str_radix(&without_prefix[4..6], 16)?;
+            Ok(RGB { red, green, blue })
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    #[test]
+    fn test_color_from() {
+        match Color::from_str("FF0000").expect("could not parse FF0000") {
+            Color::RGB(rgb) => assert_eq!(
+                rgb,
+                RGB {
+                    red: 255,
+                    green: 0,
+                    blue: 0
                 }
-                if last < txt.text.len() {
-                    current.text = txt.text[last..].to_owned();
-                    parts.push(Component::Text(current));
+            ),
+            _ => panic!("Could not parse hex color correct"),
+        }
+        match Color::from_str("#00FF00").expect("could not parse #00FF00") {
+            Color::RGB(rgb) => assert_eq!(
+                rgb,
+                RGB {
+                    red: 0,
+                    green: 255,
+                    blue: 0
                 }
-
-                let old = mem::replace(&mut txt.modifier.extra, Some(parts));
-                if let Some(old_extra) = old {
-                    if let Some(ref mut extra) = txt.modifier.extra.as_mut() {
-                        extra.extend(old_extra);
-                    }
+            ),
+            _ => panic!("Could not parse hex color correct"),
+        }
+        match Color::from_str("") {
+            Ok(_) => {}
+            Err(e) => assert_eq!(ParseColorError::InvalidLenError(0), e),
+        }
+        match Color::from_str("4343433") {
+            Ok(_) => {}
+            Err(e) => assert_eq!(ParseColorError::InvalidLenError(7), e),
+        }
+        match Color::from_str("#4343433") {
+            Ok(_) => {}
+            Err(e) => assert_eq!(ParseColorError::InvalidLenError(7), e),
+        }
+        match Color::from_str("#123456").expect("could not parse #123456") {
+            Color::RGB(rgb) => assert_eq!(
+                rgb,
+                RGB {
+                    red: 0x12,
+                    green: 0x34,
+                    blue: 0x56,
                 }
-                txt.text = "".to_owned();
-            }
+            ),
+            _ => panic!("Could not parse hex color correct"),
+        }
+
+        match Color::from_str("rEd") {
```

Comment:
The naming in these from_strs seems to be inconsistent

Reply:
no, i mean like the color in th apostrophes, like it is sometimes in caps and sometime not

Reply:
Ahh the test I change that.